docs/enhancements/auto-return-agent-to-infra-env.md

```diff
@@ -0,0 +1,140 @@
+  cluster without a BareMetalHost
+* Allow for a node to be gracefully removed from an installed cluster
+
+### Non-Goals
+
+* Automate graceful removal of a node from an existing cluster
+* Booting a node that is unresponsive or not currently part of an installed
+  cluster
+
+## Proposal
+
+When a host should be returned to the InfraEnv use the kubeconfig created when
+the cluster was installed to run an agent on the host. This agent will handle
+downloading the required artifacts and rebooting the host into the discovery
+process.
+
+### Timeline
+
+1. User triggers an action which will prepare a host to be unbound
+  * Assisted service will create a MachineConfig to run our agent on the host as
+    a systemd service.
+  * The agent will download the InfraEnv's kernel and initrd (including
+    discovery ignition) to the host's `/boot` partition.
+  * The agent will create a GRUB boot entry to boot into the discovery kernel
+    and initrd.
+2. User (or other automation) removes the node from the cluster
+  * Drain existing workloads and delete the node
+3. User unbinds the host from the cluster using the assisted-service API
+  * This triggers the agent to reboot the host into the discovery image
+4. Host is registered as `known-unbound` in the InfraEnv and is ready to be used
+   in new clusters
+
+### User Stories
+
+#### Story 1
+
+As a cluster creator I want to use the assisted service to repurpose hosts 
+from existing clusters without requiring access to the host management 
+infrastructure (BMC, hypervisor, etc).
+
+### Implementation Details/Notes/Constraints [optional]
+
+1. A way to query for URLs to the iPXE boot artifacts over the REST API should be
+implemented. Currently these URLs are only exposed through the iPXE script and
+in the InfraEnv CR, but the agent running on a node will need access to these
+artifacts.
+  * This can be added to the existing payload returned when querying an InfraEnv
+  * Alternatively this would need a new API endpoint
+    * `/v2/infra-envs/{infra_env_id}/boot-artifacts`?
+2. The agent will likely require two changes/additions
+  * A step to query the boot artifacts and write the boot entry
+  * A step to reboot the host into the new boot entry
+3. The service will need a new API to prepare a cluster for hosts to be reclaimed
+  * This needs to be separate from unbind so that the user has time to cleanly
+    remove a host from the cluster
+  * `/v2/clusters/{cluster_id}/prepare-unbind`?
+
+#### Assumptions
+
+This approach only works if the host is currently part of a working cluster
+which the assisted service can access using the kubeconfig generated during
+install.
+
+There will also need to be enough space on the `/boot` partition to download the
+initrd and kernel, but the agent can check for this during the preparation step.
+
+### Risks and Mitigations
+
+It's possible that something could go wrong during the boot process. A user 
```

Comment:
It can, this is a new agent functionality that requires new steps.
The server can parse the step response and update something on the agent so that the agent controller will be able to set the info on some condition.
Currently we update the agent conditions by looking at the validations and installation status/progress

Reply:
> Currently we update the agent conditions by looking at the validations and installation status/progress

Yeah, I think we would have some new status for this. Definitely for the time between when we tell the node to reboot and when it re-registers. I'll add details about the statemachine changes and new agent steps soon.

Reply:
With the last changes I made, the only thing we'll do is move to `UnbindingPendingUserAction` if the agent has some problem.

When the agent responds after running into an issue that response should have the error info, but currently it's not a validation so we'd have to find somewhere else to put it.

Do you think this is worth adding a new column to host for?
I guess the alternative would be to add a "preparing" state of some kind and some validation steps. Artifact access, and disk space come to mind.

What do you think?